File: src/test/java/org/projectodd/nodej/EventsTest.java
package org.projectodd.nodej;

import static org.fest.assertions.Assertions.assertThat;

import org.dynjs.runtime.JSFunction;
import org.junit.Ignore;
import org.junit.Test;

public class EventsTest extends NodejTestSupport {

    @Test
    public void testRequireEvents() {
        assertThat(eval("require('events').EventEmitter")).isInstanceOf(JSFunction.class);
    }
    
    @Test
    @Ignore
    public void testOnEvent() {
        eval("var x = 0");
        eval("var func = function() { x = x+1 }");
        eval("var events = require('events').EventEmitter");
        eval("events.prototype.on('ding', func)");
        eval("events.prototype.emit('ding')");
        assertThat(eval("x")).isEqualTo(1L);
        eval("events.prototype.emit('ding')");
        assertThat(eval("x")).isEqualTo(2L);
    }
    
    @Test
    public void testOnce() {
        eval("var x = 0");
        eval("var func = function() { x = x+1 }");
        eval("var events = require('events').EventEmitter");
        eval("events.prototype.once('ding', func)");
        eval("events.prototype.emit('ding')");
        assertThat(eval("x")).isEqualTo(1L);
        eval("events.prototype.emit('ding')");
        assertThat(eval("x")).isEqualTo(1L);
    }
    
    @Test
    public void testRemoveListener() {
        eval("var x = 0");
        eval("var func = function() { x = x+1 }");
        eval("var events = require('events').EventEmitter");
        eval("events.prototype.on('ding', func)");
        assertThat(eval("events.prototype.listeners('ding').length")).isEqualTo(1L);
        eval("events.prototype.removeListener('ding', func)");
        assertThat(eval("events.prototype.listeners('ding').length")).isEqualTo(0L);
    }

    @Test
    public void testRemoveAllListeners() {
        eval("var x = 0");
        eval("var func = function() { x = x+1 }");
        eval("var events = require('events').EventEmitter");
        eval("events.prototype.on('ding', func)");
        assertThat(eval("events.prototype.listeners('ding').length")).isEqualTo(1L);
        eval("events.prototype.removeAllListeners('ding')");
        assertThat(eval("events.prototype.listeners('ding').length")).isEqualTo(0L);
    }
}

File: src/test/resources/events/events_test.js
load("vertx.js");
load("vertx_tests.js");
EventEmitter = require('events').EventEmitter;

function testEventEmitterIsNotNull() {
  vassert.assertNotNull(EventEmitter);
  vassert.testComplete();
}

function testOnEvent() {
  var x = 0;
  var func = function() { x = x+1 }
  EventEmitter.prototype.on('ding', func);
  EventEmitter.prototype.emit('ding');
  vassert.assertEquals(1, x);
  EventEmitter.prototype.emit('ding');
  vassert.assertEquals(2, x);
  vassert.testComplete();
}

function testOnce() {
  var x = 0;
  var func = function() { x = x+1 }
  EventEmitter.prototype.once('ding', func);
  EventEmitter.prototype.emit('ding');
  vassert.assertEquals(1, x);
  EventEmitter.prototype.emit('ding');
  vassert.assertEquals(1, x);
  vassert.testComplete();
}

function testRemoveListener() {
  var x = 0;
  var func = function() { x = x+1 }
  EventEmitter.prototype.on('ding', func);
  vassert.assertEquals(1, EventEmitter.prototype.listeners('ding').length);
  EventEmitter.prototype.removeListener('ding', func);
  vassert.assertEquals(0, EventEmitter.prototype.listeners('ding').length);
  vassert.testComplete();
}

function testRemoveAllListeners() {
  var x = 0;
  var func = function() { x = x+1 }
  EventEmitter.prototype.on('ding', func);
  vassert.assertEquals(1, EventEmitter.prototype.listeners('ding').length);
  EventEmitter.prototype.removeAllListeners('ding', func);
  vassert.assertEquals(0, EventEmitter.prototype.listeners('ding').length);
  vassert.testComplete();
}

initTests(this);
